packages/lib/hooks/useTypedQuery.ts

```diff
@@ -0,0 +1,59 @@
+import { useRouter } from "next/router";
+import { z } from "zod";
+
+export function useTypedQuery<T extends z.Schema>(schema: T) {
+  type InferedSchema = z.infer<typeof schema>;
+  type SchemaKeys = keyof InferedSchema;
+  const { query: unparsedQuery, ...router } = useRouter();
+  const parsedQuery = schema.parse(unparsedQuery) as InferedSchema;
```

Comment:
You likely want to use `safeParse` here.

You don't have full control of the query params - someone can edit them manually, failing this aggressively might not be what you want